File: leetcode/590.n-ary-tree-postorder-traversal/ts/postorder.ts
class _Node {
    val: number
    children: _Node[]
    constructor(val?: number) {
        this.val = (val === undefined ? 0 : val)
        this.children = []
    }
}

const postorder = (root: _Node | null): number[] => {
    const res: number[] = []
    traversal(root, res)
    return res
}

const traversal = (root: _Node | null, res: number[]) => {
    if (root === null) return

    for (const child of root.children)
        traversal(child, res)

    res.push(root.val)

    return res
}

export default postorder

File: leetcode/590.n-ary-tree-postorder-traversal/cpp/postorder.cpp
#include <vector>

class Node
{
public:
    int val;
    std::vector<Node *> children;

    Node() {}

    Node(int _val)
    {
        val = _val;
    }

    Node(int _val, std::vector<Node *> _children)
    {
        val = _val;
        children = _children;
    }
};

class Solution
{
public:
    std::vector<int> postorder(Node *root)
    {
        std::vector<int> res;
        traversal(root, res);
        return res;
    }

private:
    void traversal(Node *root, std::vector<int> &res)
    {
        if (root == nullptr)
            return;

        for (auto n : root->children)
            traversal(n, res);

        res.push_back(root->val);
    }
};
